src/proxy/http/profiles/mod.rs

```diff
@@ -0,0 +1,282 @@
+/// middleware configurations
+///
+/// As the router's Stack is built, a destination is extracted from the stack's
+/// target and it is used to get route profiles from ` GetRoutes` implemetnation.
+///
+/// Each route uses a shared underlying concrete dst router.  The concrete dst
+/// router picks a concrete dst (NameAddr) from the profile's `dst_overrides` if
+/// they exist, or uses the router's target's addr if no `dst_overrides` exist.
+/// The concrete dst router uses the concrete dst as the target for the
+/// underlying stack.
+pub mod router;
+
+#[derive(Clone)]
+pub struct WeightedAddr {
+    pub addr: NameAddr,
+    pub weight: u32,
+}
+
+pub struct Routes {
+    pub routes: Vec<(RequestMatch, Route)>,
+    pub dst_overrides: Vec<WeightedAddr>,
+}
+
+/// Watches a destination's Routes.
+///
+/// The stream updates with all routes for the given destination. The stream
+/// never ends and cannot fail.
+pub trait GetRoutes {
+    type Stream: Stream<Item = Routes, Error = Never>;
+
+    fn get_routes(&self, dst: &NameAddr) -> Option<Self::Stream>;
+}
+
+/// Implemented by target types that may be combined with a Route.
+pub trait WithRoute {
+    type Output;
+
+    fn with_route(self, route: Route) -> Self::Output;
+}
+
+/// Implemented by target types that can have their `NameAddr` destination
+/// changed.
+pub trait WithAddr {
+    fn with_addr(self, addr: NameAddr) -> Self;
+}
+
+/// Implemented by target types that may have a `NameAddr` destination that
+/// can be discovered via `GetRoutes`.
+pub trait CanGetDestination {
+    fn get_destination(&self) -> Option<&NameAddr>;
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, Hash)]
+pub struct Route {
+    labels: Labels,
+    response_classes: ResponseClasses,
+    retries: Option<Retries>,
+    timeout: Option<Duration>,
+}
+
+#[derive(Clone, Debug)]
+pub enum RequestMatch {
+    All(Vec<RequestMatch>),
+    Any(Vec<RequestMatch>),
+    Not(Box<RequestMatch>),
+    Path(Regex),
+    Method(http::Method),
+}
+
+#[derive(Clone, Debug)]
+pub struct ResponseClass {
+    is_failure: bool,
+    match_: ResponseMatch,
+}
+
+#[derive(Clone, Default)]
+pub struct ResponseClasses(Arc<Vec<ResponseClass>>);
+
+#[derive(Clone, Debug)]
+pub enum ResponseMatch {
+    All(Vec<ResponseMatch>),
+    Any(Vec<ResponseMatch>),
+    Not(Box<ResponseMatch>),
+    Status {
+        min: http::StatusCode,
+        max: http::StatusCode,
+    },
+}
+
+#[derive(Clone, Debug)]
+pub struct Retries {
+    budget: Arc<Budget>,
+}
+
+#[derive(Clone, Default)]
+struct Labels(Arc<IndexMap<String, String>>);
+
+// === impl Route ===
+
+impl Route {
+    pub fn new<I>(label_iter: I, response_classes: Vec<ResponseClass>) -> Self
+    where
+        I: Iterator<Item = (String, String)>,
+    {
+        let labels = {
+            let mut pairs = label_iter.collect::<Vec<_>>();
+            pairs.sort_by(|(k0, _), (k1, _)| k0.cmp(k1));
```

Comment:
Is there a reason we need to sort the labels?

Reply:
I'm not sure, this block of code was simply moved in this PR.  I'd guess that we sort these so that the prometheus output is predictable and consistent.